cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 	return nil
 }
 
+func createCoreDNSConfigMap(client clientset.Interface, cm *v1.ConfigMap) error {
+	// Get the CoreDNS ConfigMap. If not present, Create a new default ConfigMap.
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Get(kubeadmconstants.CoreDNSConfigMap, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil
+		}
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Create(cm); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
+				return errors.Wrap(err, "unable to create configmap")
+			}
+		}
+	}
+	// Get the Corefile and installed CoreDNS version.
+	corefile, currentInstalledCoreDNSVersion, err := GetCoreDNSInfo(client)
+	if err != nil {
+		return errors.Wrap(err, "unable to fetch CoreDNS current installed version and ConfigMap.")
+	}
+	// Check if the current Configuration present is the default (unmodified) version.
+	// If not, try and migrate the Config.
+	if !migration.Default("", corefile) {
+		updatedCorefile, err := migration.Migrate(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile, false)
+		if err != nil {
+			return errors.Wrap(err, "unable to migrate CoreDNS Configmap")
+		}
+
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(&v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      kubeadmconstants.CoreDNSConfigMap,
+				Namespace: metav1.NamespaceSystem,
+			},
+			Data: map[string]string{
+				"Corefile":          updatedCorefile,
+				"Corefile-previous": corefile,
+			},
+		}); err != nil {
+			return errors.Wrap(err, "unable to update configmap")
+		}
+		changes, err := migration.Deprecated(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile)
+		if err != nil {
+			return errors.Wrap(err, "unable to get list of notifications")
```

Comment:
possibly use the naming "changes" in this error too?
optionally clarify more what action actually failed.